crates/trie/sparse/src/state.rs

```diff
@@ -0,0 +1,129 @@
+use crate::SparseTrie;
+use alloy_primitives::{
+    map::{HashMap, HashSet},
+    Bytes, B256,
+};
+use alloy_rlp::Decodable;
+use reth_trie::{Nibbles, TrieNode};
+
+/// Sparse state trie representing lazy-loaded Ethereum state trie.
+#[derive(Default, Debug)]
+pub struct SparseStateTrie {
+    /// Sparse account trie.
+    pub(crate) state: SparseTrie,
+    /// Sparse storage tries.
+    #[allow(dead_code)]
+    pub(crate) storages: HashMap<B256, SparseTrie>,
+    /// Collection of revealed account and storage keys.
+    #[allow(dead_code)]
+    pub(crate) revealed: HashMap<B256, HashSet<B256>>,
+}
+
+impl SparseStateTrie {
+    /// Create state trie from state trie.
+    pub fn from_state(state: SparseTrie) -> Self {
+        Self { state, ..Default::default() }
+    }
+
+    /// Returns `true` if account was already revealed.
+    pub fn is_account_revealed(&self, account: &B256) -> bool {
+        self.revealed.contains_key(account)
+    }
+
+    /// Returns `true` if storage slot for account was already revealed.
+    pub fn is_storage_slot_revealed(&self, account: &B256, slot: &B256) -> bool {
+        self.revealed.get(account).map_or(false, |slots| slots.contains(slot))
+    }
+
+    /// Reveal unknown trie paths from provided leaf path and its proof.
+    ///
+    /// # Panics
+    ///
+    /// This method panics on invalid proof if `debug_assertions` are enabled.
```

Comment:
can we return proper errors instead of doing debug assertions here?

Reply:
yup, we should migrate to proper errors